### ISSUE TITLE
Introduction

### ISSUE THREAD

**Issue by martinschroeder** [concurrent-php/task#2]
This repository contains my proposal for an async execution API in PHP. It re-uses my work on fibers with C stack switching and adds an API that should be familiar to anyone who has worked with promises or coroutines before. The fiber API is provided as a low level API for custom interruptable functions.

The most important additions are the introduction of Awaitable (which is exactly the same thing as Amp‘s Promise) and Task (comparable to a Coroutine in Amp terms). There is also a TaskScheduler that is needed to start and resume the fibers that are backing async tasks. Another important addition is the concept of a Context that allows for task-local variables (allowing for implicit value passing) and contextual error handling (which is needed when an error occurs during a continuation of an awaitable).

The static methods of the Task class are meant to be easily replaceable with new PHP keywords async and await. Introducing these keywords would improve code readability and avoid callback-style function / method calling.

**Comment by martinschroeder** [comment]
@kelunik @trowski you are quite familiar with coroutines in PHP. What do you think of this extension and the API that i am proposing? You should definitely take a look at example b.php that showcases how async tasks can be integrated into an existing event loop (which is not provided or needed by the extension in it's current state).

I am planning to add more tests and documentation and eventually want to put it on PECL if people are interested in it.

**Comment by kelunik** [comment]
@martinschroeder I wouldn't have any public API on the `Awaitable` interface. Instead, users should use `async` / `await` like so:

```
$awaitable = ...;

async function () use ($awaitable) {
    $foo = await $awaitable;
};
```

Otherwise you'll have issues with exceptions being thrown inside the callbacks.

One thing I definitely want is setting a default scheduler, so the user doesn't have to care about it at all. A user should be able to call `parallelDownload($uri...)` in an async / sync context without any issue, here it automatically runs the event loop if it's not running yet.

**Comment by kelunik** [comment]
I'd name `Task` => `GreenThread` / `UserThread`. I like contexts, but I don't think the API is ideal. IMO they should maybe be something like `ThreadLocal` in Java.

**Comment by martinschroeder** [comment]
@kelunik I implemented the minimal public API of `Awaitable` to enable custom awaitable implementations. Having just `Task` would be a little cumbersome because there would be no way to implement things like combinators (`all()`, `first()`, ...). You could basically cheat as far as `all()` is concerned by simply iterating and awaiting all tasks, but you cannot do the same thing for `first()` because you would iterate in a specific order. Another advantage is that it is very easy to integrate existing promise-based frameworks with the awaitable API, just extend from `Awaitable` and you are ready to go.

I added an example implementation of `Awaitable` to the readme file to clarify how `Context` and `Awaitable` are related. This covers the fact that no error must be thrown / bubble up from `continueWith()`.

**Comment by martinschroeder** [comment]
The `Context` API is heavily inspired by the Go language. It is immutable to avoid unpredictable modifications of the context from nested function calls. In contrast to Go i decided to pass it as an implicit value in all async executions because it allows the context to be passed through APIs that were not designed with things like async or task-local variables in mind.

Having a default scheduler would be a nice addition that would not be hard to implement. I have not implemented such a scheduler yet because i thought it would complicate unit testing code that uses tasks quite a bit as the same scheduler would be re-used across test cases.

**Comment by martinschroeder** [comment]
@kelunik context is not just something like ThreadLocal, it has variables but it also provides context-specific error handling. The missing part is cancellation, i want to implement a cancellation mechanism based on context. Cancellation would not be related to tasks, because they are just a means of providing control flow to the execution of async operations. Instead cancellation applies to async operations (awaitables) like reading from a socket or waiting for a timer to trigger. The creator of the context must be in charge of triggering cancellation (something like a cancellation token source). Contexts should not be cancellable by default, the creator of the context has to enable contextual cancellation. I have a few ideas how this might look like and will play around with it in the next days.

To sum it up: Context will be variables + error handling + (optional) cancellation.

**Comment by kelunik** [comment]
@martinschroeder As long as there's a `Deferred` implementation in the extension, custom combinators can be implemented, no? That also allows integrating with existing libraries.

> This covers the fact that no error must be thrown / bubble up from continueWith().

People won't read / obey that. What happens with thrown errors currently? I'd like to avoid that issue.

**Comment by martinschroeder** [comment]
@kelunik Adding a `Deferred` is definitely planned to reduce the boilerplate code that needs to be written for a custom `Awaitable`. I will work on it and maybe sometime later declare `Awaitable` as an interface that cannot be implemented by userland code (like `Traversable` i guess..) to get around the error handling part.

**Comment by martinschroeder** [comment]
I implemented a `Deferred` class and changed the `Awaitable` interface to throw an error if an attempt is made to implement it in a class except for `Deferred` and `Task`. I also removed related methods from `Context` because they are no longer of any use.

@kelunik This should solve the problem with error handling as it is no longer a userland concern.

**Comment by kelunik** [comment]
The method is still present on the interface, so a user defined callback throwing an exception still has the issue, no?

**Comment by martinschroeder** [comment]
Any error thrown from a (user-defined) callback is caught by eighter `Deferred` or `Task` and forwarded to the error handler of the `Context` that is bound to the object. If the error handler throws (or no error handler is set) the error will be forwarded to the parent context. The forwarding is repeated until the error is handled or the error has bubbled up to the root context. If the error is not handled by the root context it will trigger a fatal error. This is the contextual error handling mechanism (comparable to the promise error handler specified in async-interop but with contextual inheritance).

I want to keep the `Awaitable` interface like it is to allow for custom callback registrations because it is a very flexible programming model and allows for simple and efficient implementations of combinators and other things that need to work on multiple async operations at once. If you have a better idea please feel free to give a little example.

**Comment by kelunik** [comment]
I don't really like a `Context` being responsible for thread local data storage and error handling. If custom callbacks were be allowed, we wouldn't even have the problem.

Writing combinators doesn't happen frequently and removing the userland API doesn't make it harder, just less efficient. I'm totally fine with that efficiency loss if we don't need context error handling in exchange.

**Comment by martinschroeder** [comment]
I thought about your suggestions a little more and i think i have an idea how to remove the public API from `Awaitable` and still allow for efficient implementation of combinators by introducing a little helper method to the `Deferred` (or `Task`) class:
```php
public static function each(callable(Deferred, int, ?\Throwable, mixed = null) $continuation, Awaitable ...$awaitables): Awaitable { }
```
This would basically specify a map function to be applied to the given awaitables. The function would be registered as continuation and called once for every awaitable as it succeeds or fails. The first argument would be a `Deferred` that is used to succeed or fail the `Awaitable` returned from the helper method. The second argument would be the numeric index of the input awaitable. Arguments 3 and 4 are error and success value emitted by the current awaitable.

If an error is thrown from the callback function it will automatically fail the `Deferred` (and hence the result `Awaitable`, that way no contextual error handling is needed. If all awaitables have run to completion and the `Deferred` has not been resolved it will automatically fail.

The method could also accept an `array` or `iterable` as second argument instead of using varargs which would change the index parameter type to mixed. Eighter way the `continueWith()` method would be gone and `Awaitable` would just be a marker interface.

@kelunik What do you think about this? It believe it addresses your concerns about error handling (which could be dropped from `Context`) and allows for arbitrary combinators and other stuff to be implemented in a safe way.

**Comment by kelunik** [comment]
@martinschroeder I like that. :+1: 

`combine` might be a better name than `each`, because `each` doesn't make it obvious that the result is a failure if there's no explicit `resolve` on the `Deferred`.

**Comment by kelunik** [comment]
@martinschroeder `Context` seems like a bad idea to me. I've read the initial documentation block and the [linked blog post](https://blog.golang.org/context). Contexts associated with values create hidden dependencies, which are only based on arbitrary string keys. I'm probably fine with them for cancellation, but using them for data seems really bad.

**Comment by martinschroeder** [comment]
@kelunik I created a new issue #13 related to the context API.

**Comment by kelunik** [comment]
I think we can close this issue, no?

**Comment by martinschroeder** [comment]
Closing this as all the topics have eighter been addressed or (in the case of context) moved to another issue.
